tests/test_benchmark_ucx.cc

```diff
@@ -0,0 +1,481 @@
+      int server_id = global_gpu_id / local_gpu_size;
+      LOG(INFO) << "Init key for global_gpu_id " << global_gpu_id << " @ node " << server_id;
+      // Init dataScatter
+      {
+        int idx = GetKeyIndex(COMM_TYPE::DATA_SCATTER, global_session_id, global_gpu_id,
+                              global_gpu_size, num_servers);
+        CHECK_EQ(idx, server_keys_datascatter.size()) 
+            << "global_session_id: " << global_session_id
+            << " global_gpu_id: " << global_gpu_id
+            << " idx: " << idx
+            << " server_keys_datascatter.size(): " << server_keys_datascatter.size();
+        auto vals = server_vals_datascatter[idx];
+
+        ps::Key ps_key = krs[server_id].begin() + latest_key;
+        InitOneKeyThenPush(ps_key, server_keys_datascatter, server_lens, vals, len, kv, is_global_root);
+      }
+      latest_key ++;
+
+      // Init gather-scatter
+      {
+        int idx = GetKeyIndex(COMM_TYPE::GATHER, global_session_id, global_gpu_id,
+                              global_gpu_size, num_servers);
+        CHECK_EQ(idx, server_keys_gather_scatter.size());
+        auto vals = server_vals_gather_scatter[idx];
+
+        ps::Key ps_key = krs[server_id].begin() + latest_key;
+        LOG(INFO) << "Prepare key for gather-scatter " << ps_key;
+        InitOneKeyThenPush(ps_key, server_keys_gather_scatter, server_lens, vals, len, kv, is_global_root);
+      }
+      latest_key ++;
+    }
+
+    // Init dense
+    for (int server = 0; server < num_servers; server ++) {
+      {
+        int idx = GetKeyIndex(COMM_TYPE::DENSE, global_session_id, server,
+                              global_gpu_size, num_servers);
+        CHECK_EQ(idx, server_keys_dense.size());
+        auto vals = server_vals_dense[idx];
+
+        ps::Key ps_key = krs[server].begin() + latest_key;
+        InitOneKeyThenPush(ps_key, server_keys_dense, server_lens, vals, len, kv, is_global_root);
+      }
+      latest_key ++;
+    }
+  }
+
+  Postoffice::Get()->Barrier(0, ps::kWorkerGroup);
+  LOG(INFO) << "Finish setup.";
+}
+
+void RunWorker(int argc, char *argv[], KVWorker<char>* kv, int tid, int nthread) {
+  // In UCX usage mode, we have four high level APIs. All of them involes inter-node commnications
+  // that invoke ps-lite push/pull under the hood. Data is always transfered src -> dst, and push
+  // is called from src, pull called from dst. Gather and scatter share pslite keys. We call these
+  // APIs in the order of DataScatter, Gather, Scatter, DenseReduce in a worker session, and multiple
+  // worker sessions will be present in the same node. They may be in different threads in the same 
+  // process or have their own processes. Each session will be bound to a local GPU card.
+  //
+  // UCX will enable all the following src-dst memory location combinations.
+  //
+  // DataScatter: ZPush (src local CPU, dst remote GPU). A session calls a ZPush for every GPU dst on every remote node.
```

Comment:
Just to clarify
Does worker session define a separate worker performing some part of the job?
And every worker is supposed to work with a single local GPU, but all remote GPUs, right?

Reply:
(1) In our use case, all worker sessions (threads) in the same process will use the same singleton PSWorker for all communication jobs, they will annotate src / dst with the new SArray API proposed. (2) And yes, each worker will work with a single local GPU and all remote GPUs.

Reply:
Thanks.
Will every server also work with a single local GPU (so workers will push/pull to many serververs, one for each GPU)?
Or it will be a single server instance?

Reply:
We have only one server per node, working with multiple GPUs. We may move the PSServer and PSWorker in the same node to the same process, so server might not be its own process in the future (in 2 month or so) I hope this won't cause an issue for UCX context handling.

Reply:
> @brminich When I'm running this use pattern test with UCX (simply add `export DMLC_ENABLE_UCX=1` to tests/ucx_multi_node.sh and run it on two machines), I found that one of the machines (the one not colocating with ps-lite scheulder) will fail at the CHECK:
> `CHECK(addr != w_pool_.end());` in ucx_van.h L:563
> 
> Could you check if you can reproduce it and propose a fix? Thanks!
> 
> Stack trace:
> 
> ```
> #0  ps::UCXVan::GetRxBuffer (this=0x6eacb0, key=31, size=size@entry=30000000, push=<optimized out>) at ps-lite/src/./ucx_van.h:563
> #1  0x0000000000457084 in ps::UCXVan::PostRecvData (this=this@entry=0x6eacb0, meta_req=meta_req@entry=0xa0c9e8) at ps-lite/src/./ucx_van.h:682
> #2  0x0000000000457bb2 in ps::UCXVan::PollUCX (this=0x6eacb0) at ps-lite/src/./ucx_van.h:612
> #3  0x00007ffff5ce3eb0 in std::execute_native_thread_routine (__p=<optimized out>) at ../../../../../gcc-4.9.4/libstdc++-v3/src/c++11/thread.cc:84
> #4  0x00007ffff6ea51d3 in start_thread (arg=0x7fffe5476700) at pthread_create.c:309
> #5  0x00007ffff4ddfd6d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
> ```

It seems like ucx_van is assuming we never pull a key before pushing the same key, this is okay for the allreduce use case, but we implemented some APIs like GatherV in newer version of byteps.

Reply:
ok, what is the desired behavior?